kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionImpl.java

```diff
@@ -0,0 +1,156 @@
+import static io.delta.kernel.internal.util.Preconditions.checkState;
+import static io.delta.kernel.internal.util.Utils.toCloseableIterator;
+
+public class TransactionImpl implements Transaction {
+    public static final int DEFAULT_READ_VERSION = 1;
+    public static final int DEFAULT_WRITE_VERSION = 2;
+
+    private final UUID txnId = UUID.randomUUID();
+
+    private final String engineInfo;
+    private final Operation operation;
+    private final Path dataPath;
+    private final Path logPath;
+    private final Optional<StructType> newSchema;
+    private final Optional<List<String>> newPartitionColumns;
+    private final Protocol protocol;
+    private final Metadata metadata;
+    private final SnapshotImpl readSnapshot;
+
+    private boolean closed; // To avoid trying to commit the same transaction again.
+
+    public TransactionImpl(
+            Path dataPath,
+            Path logPath,
+            SnapshotImpl readSnapshot,
+            String engineInfo,
+            Operation operation,
+            Optional<StructType> newSchema,
+            Optional<List<String>> newPartitionColumns,
+            Protocol protocol,
+            Metadata metadata) {
+        this.dataPath = dataPath;
+        this.logPath = logPath;
+        this.readSnapshot = readSnapshot;
+        this.engineInfo = engineInfo;
+        this.operation = operation;
+        this.newSchema = newSchema;
+        this.newPartitionColumns = newPartitionColumns;
+        this.protocol = protocol;
+        this.metadata = metadata;
+    }
+
+    @Override
+    public Row getState(TableClient tableClient) {
+        return TransactionStateRow.of(metadata, dataPath.toString());
+    }
+
+    @Override
+    public List<String> getPartitionColumns(TableClient tableClient) {
+        return newPartitionColumns.orElseGet(() ->
+                VectorUtils.toJavaList(metadata.getPartitionColumns()));
+    }
+
+    @Override
+    public StructType getSchema(TableClient tableClient) {
+        return newSchema.orElseGet(() -> readSnapshot.getSchema(tableClient));
+    }
+
+    @Override
+    public TransactionCommitResult commit(
+            TableClient tableClient,
+            CloseableIterable<Row> dataActions) throws ConcurrentWriteException {
+        checkState(
+                !closed,
+                "Transaction is already attempted to commit. Create a new transaction.");
+        List<Row> metadataActions = new ArrayList<>();
+        metadataActions.add(createCommitInfoSingleAction(generateCommitAction()));
+        metadataActions.add(createMetadataSingleAction(metadata.toRow()));
+        metadataActions.add(createProtocolSingleAction(protocol.toRow()));
+
+
+        try (CloseableIterator<Row> stageDataIter = dataActions.iterator()) {
+            // Create a new CloseableIterator that will return the metadata actions followed by the
+            // data actions.
+            CloseableIterator<Row> stagedDataWithCommitInfo =
```

Comment:
I also think that I had commented this on the original PR. I think that this variable name makes the reader think it excludes any protocol and metadata information.

Reply:
renamed to `dataAndMetadataActions`.